### PR TITLE
ASoC: SOF: Intel/ipc4: Add support for two staged library loading

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

It was determined that the existing library loading flow cannot follow the hardware programming sequence for the DMA, which is:
DSP side sets buffer and sets the GEN bit
Host side sets the RUN bit

With the single message the current flow is:
Host sets the RUN bit
Host sends LOAD_LIBRARY
DSP side sets buffer and sets the GEN bit

The agreed solution is that we will add a new LOAD_LIBRARY_PREPARE message and by using it in host and firmware side we can have correct programming  flow:
host sends LOAD_LIBRARY_PREPARE
DSP side sets it's buffer and sets the GEN bit
Host sets the RUN bit
Host sends LOAD_LIBRARY
DSP recceives the library data.

In case the firmware does not implement this new message (returns with not supported message) then we will proceed and send only the LOAD_LIBRARY message as it was defined by legacy and to maintain backwards compatibility.

### CODE REVIEW

File: sound/soc/sof/ipc4.c [hunk 1]
```diff
@@ -99,6 +99,10 @@ static int sof_ipc4_check_reply_status(struct snd_sof_dev *sdev, u32 status)
 
 to_errno:
 	switch (status) {
+	case 2:
+	case 15:
+		ret = -EOPNOTSUPP;
+		break;
```
Comment (plbossart) [inline sound/soc/sof/ipc4.c:105]: Looks like we don't really have a consistent  way of reporting "not supported"

````
ipc3-dtrace.c:          return -EOPNOTSUPP;
ipc4-loader.c:          return -EOPNOTSUPP;
ops.h:  return -ENOTSUPP;
ops.h:  return -ENOTSUPP;
sof-client.c:   return -EOPNOTSUPP;

````

File: include/sound/sof/ipc4/header.h [hunk 2]
```diff
@@ -108,6 +108,11 @@ enum sof_ipc4_global_msg {
 
 	/* Loads library (using Code Load or HD/A Host Output DMA) */
 	SOF_IPC4_GLB_LOAD_LIBRARY,
+	/*
+	 * Prepare the host DMA channel for library loading, must be followed by
+	 * a  SOF_IPC4_GLB_LOAD_LIBRARY message as the library loading step
```
Comment (plbossart) [inline include/sound/sof/ipc4/header.h]: extra space after 'a'

File: include/sound/sof/ipc4/header.h [hunk 3]
```diff
@@ -108,6 +108,11 @@ enum sof_ipc4_global_msg {
...
+	 */
+	SOF_IPC4_GLB_LOAD_LIBRARY_PREPARE,
```
Comment (plbossart) [inline include/sound/sof/ipc4/header.h:119]: so this is the value 25, but the comment below says it's reserved. which is it?

Comment (plbossart) [reply include/sound/sof/ipc4/header.h:119]: what are the rules for adding new messages here? Is this not an ABI change, or is this considered to be ok because firmware has not been released just yet? The commit message MUST be clear on this.

File: include/sound/sof/ipc4/header.h [hunk 4]
```diff
@@ -108,6 +108,11 @@ enum sof_ipc4_global_msg {
 
 	/* Loads library (using Code Load or HD/A Host Output DMA) */
 	SOF_IPC4_GLB_LOAD_LIBRARY,
+	/*
```
Comment (plbossart) [inline include/sound/sof/ipc4/header.h:115]: typos in commit message
"it's" 
"recceives"

File: sound/soc/sof/intel/hda-loader.c [hunk 5]
```diff
@@ -546,17 +546,52 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
...
 	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
 	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
-	msg.primary |= SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID(fw_lib->id);
+
+	/*
+	 * 1st stage: SOF_IPC4_GLB_LOAD_LIBRARY_PREPARE
+	 * message with valid dma_id and invalid (0) lib_id
+	 * if the firmware does not have support for the message, we will
+	 * receive -EOPNOTSUPP. In this case we assume legacy, single step library
+	 * loading support and proceed to send the LOAD_LIBRARY message.
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c]: this goes with my question on ABI. Do we even want to support those 'legacy' firmware and why?

File: sound/soc/sof/intel/hda-loader.c [hunk 6]
```diff
@@ -546,17 +546,52 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
...
 	msg.primary = hext_stream->hstream.stream_tag - 1;
-	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_LOAD_LIBRARY);
+	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_LOAD_LIBRARY_PREPARE);
 	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
 	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
-	msg.primary |= SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID(fw_lib->id);
+
+	/*
+	 * 1st stage: SOF_IPC4_GLB_LOAD_LIBRARY_PREPARE
+	 * message with valid dma_id and invalid (0) lib_id
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c]: why do you even need to pass an invalid lib_id here?

**Review by plbossart — CHANGES_REQUESTED** [review]
a couple of cosmetic/nit-picks and a more fundamental question on ABI and definition of "legacy".

**Review by lyakh — COMMENTED** [review]
subject to addressing Pierre's comments - LGTM otherwise

File: sound/soc/sof/ipc4.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4.c:105]: I'll update the ops.h, `EOPNOTSUPP` is the preferred one.
```
$ rg EOPNOTSUPP scripts 
scripts/checkpatch.pl
4890:# Folks usually mean EOPNOTSUPP (also called ENOTSUP), when they type ENOTSUPP.
4894:                            "ENOTSUPP is not a SUSV4 error code, prefer EOPNOTSUPP\n" . $herecurr) &&
4896:                           $fixed[$fixlinenr] =~ s/\bENOTSUPP\b/EOPNOTSUPP/;
```

File: include/sound/sof/ipc4/header.h (hunk 4 above)
Comment (ujfalusi) [reply include/sound/sof/ipc4/header.h:115]: OK, rewording

File: include/sound/sof/ipc4/header.h (hunk 2 above)
Comment (ujfalusi) [reply include/sound/sof/ipc4/header.h]: Ok

File: include/sound/sof/ipc4/header.h (hunk 3 above)
Comment (ujfalusi) [reply include/sound/sof/ipc4/header.h:119]: > so this is the value 25, but the comment below says it's reserved. which is it?

It was an unused value and the firmware will be using it for this message. Dropping the comment

Comment (ujfalusi) [reply include/sound/sof/ipc4/header.h:119]: > what are the rules for adding new messages here? Is this not an ABI change, or is this considered to be ok because firmware has not been released just yet? The commit message MUST be clear on this.

As we did not have official release of IPC4 it is not yet an ABI change. We have discussed  and aligned on this internally with all teams involved with IPC4 (including the reference firmware).

`25` was reserved and not used, so firmware which does not implement this new message will reject it as unsupported.

File: sound/soc/sof/intel/hda-loader.c (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c]: True, we only send the dma_id.

File: sound/soc/sof/intel/hda-loader.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c]: I think supporting firmware which does not implement this mode of library loading can be useful and only developers will encounter them.

**Comment by ujfalusi** [comment]
Changes since v1:
- commit messages updated
- removed the note about message id 25 being reserved (as we are using it)
- clarify and cleanup the library loading flow, drop the lib_id reference from LOAD_LIBRARY_PREPARE message for example

File: sound/soc/sof/ipc4.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4.c:105]: @plbossart, https://github.com/thesofproject/linux/pull/4525 for the `ops.h`

File: include/sound/sof/ipc4/header.h (hunk 4 above)
Comment (plbossart) [reply include/sound/sof/ipc4/header.h:115]: nit-pick: DSP side sets it's buffer -> DSP side sets its buffer

File: include/sound/sof/ipc4/header.h (hunk 3 above)
Comment (plbossart) [reply include/sound/sof/ipc4/header.h:119]: I think you should really add this information in the commit message. It's very helpful to put the rules in writing for all reviewers to see, including upstream.
You added information along those lines in the next patch, but that's an Intel-specific part. We should have the information on support when the header is changed.

**Review by plbossart — APPROVED** [review]
Approved but commit messages can be improved for grammar and clarity, see below.

**Review by jsarha — DISMISSED** [review]
Code-wise I see nothing wrong. For commit messages, the message in the 3rd commit is a bit hard to follow and mostly redundant, as 4th commit message explains the same thing. Maybe remove most of the 3rd commit message, and just tell that a new message is needed. The explanation comes in the 4th message. From the 4rd I'd just drop "There is" from the beginning.  These are only suggestions, so approved.

**Comment by ujfalusi** [comment]
Changes since v2:
- commit message rewritten for the new message type introduction (SOF_IPC4_GLB_LOAD_LIBRARY_PREPARE)

**Review by plbossart — DISMISSED** [review]

File: sound/soc/sof/intel/hda-loader.c [hunk 7]
```diff
@@ -545,20 +545,58 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
...
+		 * Make sure that the FIFOS value is not 0 in SDxFIFOS register
+		 * which indicates that the firmware set the GEN bit and we can
+		 * continue to start the DMA
+		 */
+		ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_HDA_BAR,
+					sd_offset + SOF_HDA_ADSP_REG_SD_FIFOSIZE,
+					status,
+					(status & SOF_HDA_SD_FIFOSIZE_FIFOS_MASK),
+					HDA_DSP_REG_POLL_INTERVAL_US,
+					HDA_DSP_BASEFW_TIMEOUT_US);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:574]: @ujfalusi is this just a paranoia check or is it really required to check the FIFO value?

Comment (plbossart) [reply sound/soc/sof/intel/hda-loader.c:574]: ping @ujfalusi, if we want this in 6.6 we don't have much time left

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c:574]: it is paranoia check, but I feel more comfortable doing this, The firmware will set the GEN bit before it sends the reply.

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c:574]: Oh, the FIFO value check? It is paranoia and also to make sure that we only set the RUN bit after the GEN was set. It should be, but I do feel that this a good practice and provides additional point if the loading fails (like  the firmware have bug and not set the GEN bit and returns with success)

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c:574]: > ping @ujfalusi, if we want this in 6.6 we don't have much time left

I missed the comment, sorry.

File: sound/soc/sof/intel/hda-loader.c [hunk 8]
```diff
@@ -545,20 +545,58 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
...
+
+		/*
+		 * Make sure that the FIFOS value is not 0 in SDxFIFOS register
+		 * which indicates that the firmware set the GEN bit and we can
+		 * continue to start the DMA
+		 */
+		ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_HDA_BAR,
+					sd_offset + SOF_HDA_ADSP_REG_SD_FIFOSIZE,
+					status,
+					(status & SOF_HDA_SD_FIFOSIZE_FIFOS_MASK),
```
Comment (lyakh) [inline sound/soc/sof/intel/hda-loader.c]: superfluous parentheses

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-loader.c]: OK, fixing it, for some reason checkpatch was fine with it.

**Comment by ujfalusi** [comment]
Changes since v3:
- remove unneeded parenthesis in snd_sof_dsp_read_poll_timeout() call

File: sound/soc/sof/intel/hda-loader.c (hunk 8 above)
Comment (lyakh) [reply sound/soc/sof/intel/hda-loader.c]: oh, checkpatch is perfectly happy to swallow as many parentheses as you want to throw at it `x = (((((0)))));`

**Review by plbossart — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Review by jsarha — APPROVED** [review]

**Review by lyakh — APPROVED** [review]

**Comment by mengdonglin** [comment]
This is kernel part for loading a module library. The FW part has been merged https://github.com/thesofproject/sof/pull/7986

**Comment by ujfalusi** [comment]
> This is kernel part for loading a module library. The FW part has been merged [thesofproject/sof#7986](https://github.com/thesofproject/sof/pull/7986)

@mengdonglin, that PR was the cleanup and fixes, the two staged one is not yet submitted.

**Comment by plbossart** [comment]
@ranj063 can you review this one please?

**Review by ranj063 — APPROVED** [review]

**Comment by ranj063** [comment]
@ujfalusi there's a genuine checkpatch warning. Maybe you can submit a fixup?

**Comment by ujfalusi** [comment]
> @ujfalusi there's a genuine checkpatch warning. Maybe you can submit a fixup?

There are two warnings, one is for 'hda' (misspelled had?) the other is alignment, but that I cannot fix unless we prefer long line warning. Almost all uses of `snd_sof_dsp_read_poll_timeout()` are like this.

**Review by aiChaoSONG — APPROVED** [review]

**Comment by aiChaoSONG** [comment]
> > @ujfalusi there's a genuine checkpatch warning. Maybe you can submit a fixup?
> 
> There are two warnings, one is for 'hda' (misspelled had?) the other is alignment, but that I cannot fix unless we prefer long line warning. Almost all uses of `snd_sof_dsp_read_poll_timeout()` are like this.

Guess what, the checkpatch issue could be fixed without the long line warning.
